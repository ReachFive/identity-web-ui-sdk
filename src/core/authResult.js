import { parseJwtTokenPayload } from '../helpers/jwt'
import { logError } from '../helpers/logger'

export function enrichAuthResult(response) {
    if (response.idToken) {
        try {
            const idTokenPayload = parseJwtTokenPayload(response.idToken);

            return {
                ...response,
                idTokenPayload
            };
        } catch (e) {
            logError('id token parsing error: ' + e);
        }
    }

    return response;
}

export function isAuthResult(value) {
    return value && (value.accessToken || value.idToken);
}

export function isErrorResponse(value) {
    return value && value.error;
}
