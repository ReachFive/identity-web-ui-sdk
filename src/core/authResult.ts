import { AuthResult, ErrorResponse } from '@reachfive/identity-core';

import { parseJwtTokenPayload } from '../helpers/jwt';
import { logError } from '../helpers/logger';

export function enrichAuthResult(response: AuthResult) {
    if (response.idToken) {
        try {
            const idTokenPayload = parseJwtTokenPayload(response.idToken);

            return {
                ...response,
                idTokenPayload,
            };
        } catch (e) {
            logError('id token parsing error: ' + e);
        }
    }

    return response;
}

export function isAuthResult(value: unknown): value is AuthResult {
    return (
        value !== null &&
        typeof value === 'object' &&
        ('accessToken' in value || 'idToken' in value)
    );
}

export function isErrorResponse(value: unknown): value is ErrorResponse {
    return value !== null && typeof value === 'object' && 'error' in value;
}
