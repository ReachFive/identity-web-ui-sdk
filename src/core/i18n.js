import isPlainObject from 'lodash-es/isPlainObject';

export default function resolveI18n(defaultMessages, messages = {}) {
    const mergedMessages = {
        ...defaultMessages,
        ...flattenObject(messages)
    };

    return (key, params) => {
        const template = mergedMessages[key] || key;

        return params
            ? Object.keys(params).reduce((acc, param) => acc.replace(`{${param}}`, params[param]), template)
            : template;
    };
}

function flattenObject(object, prefix = []) {
    return isPlainObject(object)
        ? Object.keys(object).reduce((acc, key) => ({
            ...acc,
            ...flattenObject(object[key], [...prefix, key])
        }), {})
        : ({ [prefix.join('.')]: object });
}
