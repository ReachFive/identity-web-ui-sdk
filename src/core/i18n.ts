export type I18nMessages = Record<string, string>

export type I18nNestedMessages = Record<string, string | I18nMessages>

export type I18nMessageParams = Record<string, unknown>

export type I18nResolver = (key: string, params?: I18nMessageParams, fallback?: (params?: I18nMessageParams) => string) => string

export function resolveI18n<T extends (Record<string, T> | string)>(defaultMessages: I18nMessages = {}, messages: I18nNestedMessages = {}): I18nResolver {
    const mergedMessages: I18nMessages = {
        ...defaultMessages,
        ...flattenObject(messages)
    };

    return (key: string, params?: I18nMessageParams, fallback?: (params?: I18nMessageParams) => string) => {
        const template = mergedMessages[key] ?? fallback?.(params) ?? key;

        return params
            ? Object.keys(params).reduce((acc, param) => acc.replace(`{${param}}`, params[param] as string), template)
            : template;
    };
}

export default resolveI18n

function flattenObject(object: I18nNestedMessages | string, prefix: string[] = []): Record<string, string> {
    return typeof object === "object" && object !== null
        ? Object.keys(object).reduce((acc, key) => ({
            ...acc,
            ...flattenObject(object[key], [...prefix, key])
        }), {})
        : ({ [prefix.join('.')]: object });
}
