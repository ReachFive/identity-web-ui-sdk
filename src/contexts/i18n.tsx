import React, { ComponentType, PropsWithChildren, useMemo } from 'react';

import { I18nMessages, I18nNestedMessages, I18nResolver, resolveI18n } from '../core/i18n';

export interface Props {
  defaultMessages?: I18nMessages
  messages?: I18nNestedMessages
}

export const I18nContext = React.createContext<I18nResolver | undefined>(undefined);

export function useI18n(): I18nResolver {
  const context = React.useContext(I18nContext);
  if (!context) {
    throw new Error('No I18nContext provided');
  }

  return context;
}

export interface I18nProps {
  i18n: I18nResolver
}

export type WithI18n<P> = P & I18nProps

export function withI18n<T extends I18nProps = I18nProps>(WrappedComponent: ComponentType<T>) {
  const displayName = WrappedComponent.displayName || WrappedComponent.name || "Component";

  const ComponentWithI18n = (props: Omit<T, keyof I18nProps>) => {
    const i18n = useI18n();
    return <WrappedComponent {...({ i18n })} {...(props as T)} />;
  };

  ComponentWithI18n.displayName = `withI18n(${displayName})`;

  return ComponentWithI18n;
}

export function I18nProvider({ children, defaultMessages, messages }: PropsWithChildren<Props>): JSX.Element | null {
  const resolver = useMemo(() => resolveI18n(defaultMessages, messages), [defaultMessages, messages])
  return <I18nContext.Provider value={resolver}>{children}</I18nContext.Provider>
}
