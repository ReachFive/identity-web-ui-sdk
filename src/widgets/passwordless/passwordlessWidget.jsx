import React from 'react';

import { email } from '../../core/validation';

import { createMultiViewWidget } from '../../components/widget/widget';
import { Info, Intro, Separator } from '../../components/miscComponent';

import { createForm } from '../../components/form/formComponent';
import { simpleField } from '../../components/form/fields/simpleField';
import phoneNumberField from '../../components/form/fields/phoneNumberField';
import SocialButtons from '../../components/form/socialButtonsComponent';

const EmailInputForm = createForm({
    prefix: 'r5-passwordless-',
    fields: [
        simpleField({
            key: 'email',
            label: 'email',
            type: 'email',
            validator: email
        })
    ]
});

const phoneNumberInputForm = config => createForm({
    prefix: 'r5-passwordless-sms-',
    fields: [phoneNumberField({ required: true }, config)]
});

const VerificationCodeInputForm = createForm({
    prefix: 'r5-passwordless-sms-',
    fields: [
        simpleField({
            key: 'verification_code',
            label: 'verificationCode',
            type: 'text'
        })
    ]
});

class MainView extends React.Component {
    handleSubmit = data => this.props.apiClient.startPasswordless(data, this.props.auth).then(_ => data);

    handleSuccess = data => {
        return data.email
            ? this.props.goTo('emailSent')
            : this.props.goTo('verificationCode', data);
    };

    render() {
        const { i18n, showSocialLogins, socialProviders, showIntro, authType } = this.props;
        const isEmail = authType === 'magic_link' || authType === '' || !authType;
        const PhoneNumberInputForm = phoneNumberInputForm(this.props.config);

        return <div>
            {showSocialLogins && socialProviders && socialProviders.length > 0 && (
                <SocialButtons providers={socialProviders} auth={this.props.auth} />
            )}
            {showSocialLogins && socialProviders && socialProviders.length > 0 && (
                <Separator text={i18n('or')} />
            )}
            {isEmail && showIntro && <Intro>{i18n('passwordless.intro')}</Intro>}
            {isEmail && <EmailInputForm handler={this.handleSubmit} onSuccess={this.handleSuccess} />}
            {!isEmail && showIntro && <Intro>{i18n('passwordless.sms.intro')}</Intro>}
            {!isEmail && <PhoneNumberInputForm handler={this.handleSubmit} onSuccess={this.handleSuccess} />}
        </div>;
    }
}

class VerificationCodeView extends React.Component {
    handleSubmit = data => {
        const { apiClient, auth, phoneNumber } = this.props;

        return apiClient.verifyPasswordless({ phoneNumber, ...data }, auth);
    };

    render() {
        return <div>
            <Info>{this.props.i18n('passwordless.sms.verification.intro')}</Info>
            <VerificationCodeInputForm handler={this.handleSubmit} />
        </div>;
    }
}
const EmailSentView = ({ i18n }) => <Info>{i18n('passwordless.emailSent')}</Info>;

export default createMultiViewWidget({
    initialView: 'main',
    views: {
        main: MainView,
        emailSent: EmailSentView,
        verificationCode: VerificationCodeView
    },
    prepare: (options, { config }) => ({
        socialProviders: config.socialProviders,
        auth: {},
        showSocialLogins: false,
        showIntro: true,
        authType: 'magic_link',
        ...options
    })
})
