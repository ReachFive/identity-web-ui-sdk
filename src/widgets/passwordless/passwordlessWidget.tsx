import React, { ComponentProps, useLayoutEffect } from 'react';

import { AuthOptions, AuthResult, SingleFactorPasswordlessParams } from '@reachfive/identity-core';

import { CaptchaProvider, WithCaptchaProps, type WithCaptchaToken } from '../../components/captcha';
import phoneNumberField, {
    type PhoneNumberOptions,
} from '../../components/form/fields/phoneNumberField';
import { simpleField } from '../../components/form/fields/simpleField';
import { createForm } from '../../components/form/formComponent';
import { SocialButtons } from '../../components/form/socialButtonsComponent';
import { Info, Intro, Separator } from '../../components/miscComponent';
import { importGoogleRecaptchaScript } from '../../components/reCaptcha';
import { createMultiViewWidget } from '../../components/widget/widget';
import { useI18n } from '../../contexts/i18n';
import { useReachfive } from '../../contexts/reachfive';
import { useRouting } from '../../contexts/routing';
import { email } from '../../core/validation';

import type { Config, OnError, OnSuccess, Prettify } from '../../types';

type EmailFormData = { email: string; captchaToken?: string };

const EmailInputForm = createForm<EmailFormData>({
    prefix: 'r5-passwordless-',
    fields: [
        simpleField({
            key: 'email',
            label: 'email',
            type: 'email',
            validator: email,
        }),
    ],
});

type PhoneNumberFormData = { phoneNumber: string; captchaToken?: string };

const phoneNumberInputForm = (config: Config) =>
    createForm<PhoneNumberFormData, { phoneNumberOptions?: PhoneNumberOptions }>({
        prefix: 'r5-passwordless-sms-',
        fields: ({ phoneNumberOptions }) => [
            phoneNumberField(
                {
                    required: true,
                    ...phoneNumberOptions,
                },
                config
            ),
        ],
    });

type VerificationCodeFormData = { verificationCode: string };

const VerificationCodeInputForm = createForm<VerificationCodeFormData>({
    prefix: 'r5-passwordless-sms-',
    fields: [
        simpleField({
            key: 'verification_code',
            label: 'verificationCode',
            type: 'text',
        }),
    ],
});

interface MainViewProps {
    /**
     * List of authentication options
     */
    auth?: AuthOptions;
    /**
     * The passwordless auth type (`magic_link` or `sms`).
     * @default "magic_link"
     */
    authType?: SingleFactorPasswordlessParams['authType'];
    /**
     * Show the introduction text.
     * @default true
     */
    showIntro?: boolean;
    /**
     * Show the social login buttons.
     * @default false
     */
    showSocialLogins?: boolean;
    /**
     * Lists the available social providers. This is an array of strings.
     *
     * Tip:  If you pass an empty array, social providers will not be displayed.
     */
    socialProviders?: string[];
    /**
     * Phone number field options.
     */
    phoneNumberOptions?: PhoneNumberOptions;
    /**
     * Callback function called when the request has succeed.
     */
    onSuccess?: OnSuccess;
    /**
     * Callback function called when the request has failed.
     */
    onError?: OnError;
}

const MainView = ({
    auth,
    authType = 'magic_link',
    recaptcha_enabled = false,
    recaptcha_site_key,
    captchaFoxEnabled = false,
    captchaFoxMode = 'hidden',
    captchaFoxSiteKey,
    showIntro = true,
    showSocialLogins = false,
    socialProviders,
    phoneNumberOptions,
    onError = (() => {}) as OnError,
    onSuccess = (() => {}) as OnSuccess,
}: WithCaptchaProps<MainViewProps>) => {
    const { client: coreClient, config } = useReachfive();
    const i18n = useI18n();
    const { goTo } = useRouting();

    useLayoutEffect(() => {
        importGoogleRecaptchaScript(recaptcha_site_key);
    }, [recaptcha_site_key]);

    const callback = (data: WithCaptchaToken<EmailFormData | PhoneNumberFormData>) =>
        coreClient
            .startPasswordless(
                {
                    authType,
                    ...data,
                },
                auth
            )
            .then(() => data);

    const handleSuccess = (data: EmailFormData | PhoneNumberFormData) => {
        onSuccess({ name: 'otp_sent', authType });
        if ('email' in data) {
            goTo('emailSent');
        } else {
            goTo<VerificationCodeViewState>('verificationCode', data);
        }
    };

    const isEmail = authType === 'magic_link';
    const PhoneNumberInputForm = phoneNumberInputForm(config);

    return (
        <div>
            {showSocialLogins && socialProviders && socialProviders.length > 0 && (
                <SocialButtons
                    providers={socialProviders}
                    auth={auth}
                    onSuccess={onSuccess}
                    onError={onError}
                />
            )}
            {showSocialLogins && socialProviders && socialProviders.length > 0 && (
                <Separator text={i18n('or')} />
            )}
            {isEmail && showIntro && <Intro>{i18n('passwordless.intro')}</Intro>}
            {isEmail && (
                <CaptchaProvider
                    recaptcha_enabled={recaptcha_enabled}
                    recaptcha_site_key={recaptcha_site_key}
                    captchaFoxEnabled={captchaFoxEnabled}
                    captchaFoxSiteKey={captchaFoxSiteKey}
                    captchaFoxMode={captchaFoxMode}
                    action="passwordless_email"
                >
                    <EmailInputForm
                        handler={callback}
                        onSuccess={handleSuccess}
                        onError={onError}
                    />
                </CaptchaProvider>
            )}
            {!isEmail && showIntro && <Intro>{i18n('passwordless.sms.intro')}</Intro>}
            {!isEmail && (
                <CaptchaProvider
                    recaptcha_enabled={recaptcha_enabled}
                    recaptcha_site_key={recaptcha_site_key}
                    captchaFoxEnabled={captchaFoxEnabled}
                    captchaFoxSiteKey={captchaFoxSiteKey}
                    captchaFoxMode={captchaFoxMode}
                    action="passwordless_phone"
                >
                    <PhoneNumberInputForm
                        handler={callback}
                        onSuccess={handleSuccess}
                        onError={onError}
                        phoneNumberOptions={phoneNumberOptions}
                    />
                </CaptchaProvider>
            )}
        </div>
    );
};

interface VerificationCodeViewProps {
    /**
     * The passwordless auth type (`magic_link` or `sms`).
     * @default "magic_link"
     */
    authType?: SingleFactorPasswordlessParams['authType'];
    /**
     * Callback function called when the request has succeed.
     */
    onSuccess?: OnSuccess;
    /**
     * Callback function called when the request has failed.
     */
    onError?: OnError;
}

type VerificationCodeViewState = {
    phoneNumber: string;
};

const VerificationCodeView = ({
    authType = 'magic_link',
    recaptcha_enabled = false,
    recaptcha_site_key,
    captchaFoxEnabled = false,
    captchaFoxMode = 'hidden',
    captchaFoxSiteKey,
    onSuccess = (() => {}) as OnSuccess,
    onError = (() => {}) as OnError,
}: WithCaptchaProps<VerificationCodeViewProps>) => {
    const { client: coreClient } = useReachfive();
    const i18n = useI18n();
    const { params } = useRouting();
    const { phoneNumber } = params as VerificationCodeViewState;

    const handleSubmit = (data: WithCaptchaToken<VerificationCodeFormData>) => {
        return coreClient
            .verifyPasswordless({
                authType,
                phoneNumber,
                ...data,
            })
            .then(result => {
                if (AuthResult.isAuthResult(result)) {
                    onSuccess({
                        name: 'login',
                        authResult: result,
                        authType,
                        identifierType: 'phone_number',
                    });
                } else {
                    onError();
                }
            })
            .catch(onError);
    };

    return (
        <div>
            <CaptchaProvider
                recaptcha_enabled={recaptcha_enabled}
                recaptcha_site_key={recaptcha_site_key}
                captchaFoxEnabled={captchaFoxEnabled}
                captchaFoxSiteKey={captchaFoxSiteKey}
                captchaFoxMode={captchaFoxMode}
                action="verify_passwordless_sms"
            >
                <Info>{i18n('passwordless.sms.verification.intro')}</Info>
                <VerificationCodeInputForm handler={handleSubmit} onError={onError} />
            </CaptchaProvider>
        </div>
    );
};

const EmailSentView = () => {
    const i18n = useI18n();
    return <Info>{i18n('passwordless.emailSent')}</Info>;
};

export type PasswordlessWidgetProps = Prettify<
    ComponentProps<typeof MainView> & ComponentProps<typeof VerificationCodeView>
>;

export default createMultiViewWidget<PasswordlessWidgetProps>({
    initialView: 'main',
    views: {
        main: MainView,
        emailSent: EmailSentView,
        verificationCode: VerificationCodeView,
    },
    prepare: (options, { config }) => ({
        socialProviders: config.socialProviders,
        ...options,
    }),
});
