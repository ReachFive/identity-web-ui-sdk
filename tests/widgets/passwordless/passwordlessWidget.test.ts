/**
 * @jest-environment jsdom
 */

import { beforeEach, describe, expect, jest, test } from '@jest/globals';
import '@testing-library/jest-dom/jest-globals';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import 'jest-styled-components';

import { type Client } from '@reachfive/identity-core';

import { type I18nMessages } from '../../../src/core/i18n';
import type { Config, OnError, OnSuccess } from '../../../src/types';

import passwordlessWidget from '../../../src/widgets/passwordless/passwordlessWidget';

const defaultConfig: Config = {
    clientId: 'local',
    domain: 'local.reach5.net',
    sso: false,
    sms: false,
    webAuthn: false,
    language: 'fr',
    pkceEnforced: false,
    isPublic: true,
    socialProviders: ['facebook', 'google'],
    customFields: [],
    resourceBaseUrl: 'http://localhost',
    mfaSmsEnabled: false,
    mfaEmailEnabled: false,
    rbaEnabled: false,
    consentsVersions: {},
    passwordPolicy: {
        minLength: 8,
        minStrength: 2,
        allowUpdateWithAccessTokenOnly: true,
    },
};

const defaultI18n: I18nMessages = {};

describe('Snapshot', () => {
    const generateSnapshot =
        (options: Parameters<typeof passwordlessWidget>[0] = {}, config: Partial<Config> = {}) =>
        async () => {
            // @ts-expect-error partial Client
            const apiClient: Client = {};

            const widget = await passwordlessWidget(options, {
                config: { ...defaultConfig, ...config },
                apiClient,
                defaultI18n,
            });

            await waitFor(async () => {
                const { container } = await render(widget);
                expect(container).toMatchSnapshot();
            });
        };

    describe('passwordless', () => {
        test('default', generateSnapshot());

        test('no intro', generateSnapshot({ showIntro: false }));

        test('sms', generateSnapshot({ authType: 'sms' }));
    });
});

describe('DOM testing', () => {
    const startPasswordless = jest.fn<Client['startPasswordless']>();
    const verifyPasswordless = jest.fn<Client['verifyPasswordless']>();

    const onError = jest.fn<OnError>();
    const onSuccess = jest.fn<OnSuccess>();

    beforeEach(() => {
        startPasswordless.mockClear();
        verifyPasswordless.mockClear();
        onError.mockClear();
        onSuccess.mockClear();
    });

    const generateComponent = async (
        options: Parameters<typeof passwordlessWidget>[0] = {},
        config: Partial<Config> = {}
    ) => {
        // @ts-expect-error partial Client
        const apiClient: Client = {
            startPasswordless,
            verifyPasswordless,
        };

        const result = await passwordlessWidget(
            { onError, onSuccess, ...options },
            { config: { ...defaultConfig, ...config }, apiClient, defaultI18n }
        );

        return waitFor(async () => render(result));
    };

    describe('passwordless', () => {
        test('default', async () => {
            expect.assertions(8);

            const user = userEvent.setup();

            startPasswordless.mockResolvedValue({});

            await generateComponent();

            // Intro
            expect(screen.queryByText('passwordless.intro')).toBeInTheDocument();

            // Label
            expect(screen.queryByLabelText('email')).toBeInTheDocument();

            // Input email
            const emailInput = screen.getByTestId('email');
            expect(emailInput).toBeInTheDocument();

            // Form button
            const submitBtn = screen.getByTestId('submit');
            expect(submitBtn).toHaveTextContent('send');

            await user.type(emailInput, 'alice@reach5.co');
            await user.click(submitBtn);

            expect(startPasswordless).toBeCalledWith(
                expect.objectContaining({
                    authType: 'magic_link',
                    email: 'alice@reach5.co',
                }),
                undefined // auth
            );

            expect(screen.queryByText('passwordless.emailSent')).toBeInTheDocument();

            expect(onSuccess).toBeCalledWith(
                expect.objectContaining({
                    authType: 'magic_link',
                    name: 'otp_sent',
                })
            );
            expect(onError).not.toBeCalled();
        });

        test('no intro', async () => {
            expect.assertions(2);
            await generateComponent({ showIntro: false });

            // Intro
            expect(screen.queryByText('passwordless.intro')).not.toBeInTheDocument();
            expect(screen.queryByText('passwordless.sms.intro')).not.toBeInTheDocument();
        });

        test('by phone number', async () => {
            expect.assertions(10);

            const user = userEvent.setup();

            startPasswordless.mockResolvedValue({
                challengeId: 'azerty',
            });

            verifyPasswordless.mockResolvedValue({
                accessToken: 'abcd1234',
            });

            await generateComponent({ authType: 'sms' });

            // Intro
            expect(screen.queryByText('passwordless.sms.intro')).toBeInTheDocument();

            // Label
            expect(screen.queryByLabelText('phoneNumber')).toBeInTheDocument();

            // Input phone number
            const phoneNumberInput = screen.getByTestId('phone_number');
            expect(phoneNumberInput).toBeInTheDocument();

            // Form button
            const submitBtn = screen.getByTestId('submit');
            expect(submitBtn).toHaveTextContent('send');

            await user.type(phoneNumberInput, '+33612345678');
            await user.click(submitBtn);

            expect(startPasswordless).toBeCalledWith(
                expect.objectContaining({
                    authType: 'sms',
                    phoneNumber: '+33612345678',
                }),
                undefined // auth
            );

            const verificationCodeInput = screen.getByTestId('verification_code');
            expect(verificationCodeInput).toBeInTheDocument();

            const submitCodeBtn = screen.getByTestId('submit');
            expect(submitCodeBtn).toHaveTextContent('send');

            await user.type(verificationCodeInput, '123456');
            await user.click(submitCodeBtn);

            expect(verifyPasswordless).toBeCalledWith(
                expect.objectContaining({
                    authType: 'sms',
                    phoneNumber: '+33612345678',
                    verificationCode: '123456',
                })
            );

            expect(onSuccess).toBeCalledWith(
                expect.objectContaining({
                    authResult: expect.objectContaining({ accessToken: 'abcd1234' }),
                    name: 'login',
                })
            );
            expect(onError).not.toBeCalled();
        });

        test('api failure', async () => {
            const user = userEvent.setup();

            startPasswordless.mockRejectedValue('Unexpected error');

            await generateComponent();

            const emailInput = screen.getByTestId('email');
            const submitBtn = screen.getByTestId('submit');

            await user.type(emailInput, 'alice@reach5.co');
            await user.click(submitBtn);

            expect(startPasswordless).toBeCalled();

            expect(onSuccess).not.toBeCalled();
            expect(onError).toBeCalledWith('Unexpected error');
        });

        test('api failure on code verification', async () => {
            const user = userEvent.setup();

            startPasswordless.mockResolvedValue({
                challengeId: 'azerty',
            });

            verifyPasswordless.mockRejectedValue('Unexpected error');

            await generateComponent({ authType: 'sms' });

            const phoneNumberInput = screen.getByTestId('phone_number');
            const submitBtn = screen.getByTestId('submit');

            await user.type(phoneNumberInput, '+33612345678');
            await user.click(submitBtn);

            expect(startPasswordless).toBeCalled();

            const verificationCodeInput = screen.getByTestId('verification_code');
            const submitCodeBtn = screen.getByTestId('submit');

            await user.type(verificationCodeInput, '123456');
            await user.click(submitCodeBtn);

            expect(verifyPasswordless).toBeCalled();

            expect(onSuccess).not.toBeCalled();
            expect(onError).toBeCalledWith('Unexpected error');
        });
    });
});
