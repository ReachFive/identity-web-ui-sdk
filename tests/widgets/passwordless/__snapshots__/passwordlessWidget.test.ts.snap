// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot passwordless default 1`] = `
.c1 {
  text-align: center;
  margin-bottom: 10px;
  color: #495057;
}

.c0 {
  font-size: 14px;
  -webkit-transition: -webkit-transform 400ms ease, opacity 400ms ease;
  -webkit-transition: transform 400ms ease, opacity 400ms ease;
  transition: transform 400ms ease, opacity 400ms ease;
  opacity: 0;
  padding: 20px;
  border-radius: 3px;
  background-color: #ffffff;
  max-width: 400px;
  box-sizing: border-box;
  margin: 0 auto;
}

.c6 {
  display: block;
  width: 100%;
  box-sizing: border-box;
  text-align: center;
  font-weight: bold;
  vertical-align: middle;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  touch-action: manipulation;
  cursor: pointer;
  color: #ffffff;
  background-color: #229955;
  border: 1px solid #229955;
  padding: 9px 12px;
  font-size: 14px;
  line-height: 1.428571429;
  border-radius: 3px;
  -webkit-transition: all .15s ease-in-out;
  transition: all .15s ease-in-out;
}

.c6:focus {
  outline: 0;
  box-shadow: 0 0 0 3px rgba(34,153,85,0.5);
}

.c6:hover,
.c6:active {
  background-color: #1b7842;
  border-color: #1b7842;
}

.c6[disabled] {
  opacity: .65;
}

.c2 {
  position: relative;
}

.c4 {
  color: #495057;
  margin-bottom: 5px;
  display: none;
}

.c3 {
  margin-bottom: 10px;
}

.c3 > label::after {
  content: "\\A0*";
  color: #dc4e41;
}

.c5 {
  display: block;
  width: 100%;
  box-sizing: border-box;
  font-size: 14px;
  line-height: 1.428571429;
  color: #495057;
  border-radius: 3px;
  border-width: 1px;
  border-style: solid;
  border-color: #ced4da;
  background-color: #fff;
  background-image: none;
  padding: 9px 12px;
  -webkit-transition: border-color ease-in-out .15s, box-shadow ease-in-out .15s;
  transition: border-color ease-in-out .15s, box-shadow ease-in-out .15s;
  box-shadow: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c5:focus {
  border-color: #5fdb94;
  box-shadow: 0 0 0 3px rgba(95,219,148,0.5);
  outline: 0;
}

.c5:disabled,
.c5[readonly] {
  background-color: #e9ecef;
}

.c5::-webkit-input-placeholder {
  color: #868e96;
}

.c5::-moz-placeholder {
  color: #868e96;
}

.c5:-ms-input-placeholder {
  color: #868e96;
}

.c5::placeholder {
  color: #868e96;
}

<div>
  <div
    class="c0"
  >
    <div>
      <div
        class="c1"
      >
        passwordless.intro
      </div>
      <form
        class="c2"
        novalidate=""
      >
        <div
          class="c3"
          required=""
        >
          <label
            class="c4"
            for="email"
          >
            email
          </label>
          <input
            class="c5"
            data-testid="email"
            id="email"
            name="email"
            placeholder="email"
            required=""
            title="email"
            type="email"
            value=""
          />
        </div>
        <button
          class="c6"
          data-testid="submit"
          type="submit"
        >
          send
        </button>
      </form>
    </div>
  </div>
</div>
`;

exports[`Snapshot passwordless no intro 1`] = `
.c0 {
  font-size: 14px;
  -webkit-transition: -webkit-transform 400ms ease, opacity 400ms ease;
  -webkit-transition: transform 400ms ease, opacity 400ms ease;
  transition: transform 400ms ease, opacity 400ms ease;
  opacity: 0;
  padding: 20px;
  border-radius: 3px;
  background-color: #ffffff;
  max-width: 400px;
  box-sizing: border-box;
  margin: 0 auto;
}

.c5 {
  display: block;
  width: 100%;
  box-sizing: border-box;
  text-align: center;
  font-weight: bold;
  vertical-align: middle;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  touch-action: manipulation;
  cursor: pointer;
  color: #ffffff;
  background-color: #229955;
  border: 1px solid #229955;
  padding: 9px 12px;
  font-size: 14px;
  line-height: 1.428571429;
  border-radius: 3px;
  -webkit-transition: all .15s ease-in-out;
  transition: all .15s ease-in-out;
}

.c5:focus {
  outline: 0;
  box-shadow: 0 0 0 3px rgba(34,153,85,0.5);
}

.c5:hover,
.c5:active {
  background-color: #1b7842;
  border-color: #1b7842;
}

.c5[disabled] {
  opacity: .65;
}

.c1 {
  position: relative;
}

.c3 {
  color: #495057;
  margin-bottom: 5px;
  display: none;
}

.c2 {
  margin-bottom: 10px;
}

.c2 > label::after {
  content: "\\A0*";
  color: #dc4e41;
}

.c4 {
  display: block;
  width: 100%;
  box-sizing: border-box;
  font-size: 14px;
  line-height: 1.428571429;
  color: #495057;
  border-radius: 3px;
  border-width: 1px;
  border-style: solid;
  border-color: #ced4da;
  background-color: #fff;
  background-image: none;
  padding: 9px 12px;
  -webkit-transition: border-color ease-in-out .15s, box-shadow ease-in-out .15s;
  transition: border-color ease-in-out .15s, box-shadow ease-in-out .15s;
  box-shadow: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c4:focus {
  border-color: #5fdb94;
  box-shadow: 0 0 0 3px rgba(95,219,148,0.5);
  outline: 0;
}

.c4:disabled,
.c4[readonly] {
  background-color: #e9ecef;
}

.c4::-webkit-input-placeholder {
  color: #868e96;
}

.c4::-moz-placeholder {
  color: #868e96;
}

.c4:-ms-input-placeholder {
  color: #868e96;
}

.c4::placeholder {
  color: #868e96;
}

<div>
  <div
    class="c0"
  >
    <div>
      <form
        class="c1"
        novalidate=""
      >
        <div
          class="c2"
          required=""
        >
          <label
            class="c3"
            for="email"
          >
            email
          </label>
          <input
            class="c4"
            data-testid="email"
            id="email"
            name="email"
            placeholder="email"
            required=""
            title="email"
            type="email"
            value=""
          />
        </div>
        <button
          class="c5"
          data-testid="submit"
          type="submit"
        >
          send
        </button>
      </form>
    </div>
  </div>
</div>
`;

exports[`Snapshot passwordless sms 1`] = `
.c1 {
  text-align: center;
  margin-bottom: 10px;
  color: #495057;
}

.c0 {
  font-size: 14px;
  -webkit-transition: -webkit-transform 400ms ease, opacity 400ms ease;
  -webkit-transition: transform 400ms ease, opacity 400ms ease;
  transition: transform 400ms ease, opacity 400ms ease;
  opacity: 0;
  padding: 20px;
  border-radius: 3px;
  background-color: #ffffff;
  max-width: 400px;
  box-sizing: border-box;
  margin: 0 auto;
}

.c6 {
  display: block;
  width: 100%;
  box-sizing: border-box;
  text-align: center;
  font-weight: bold;
  vertical-align: middle;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  touch-action: manipulation;
  cursor: pointer;
  color: #ffffff;
  background-color: #229955;
  border: 1px solid #229955;
  padding: 9px 12px;
  font-size: 14px;
  line-height: 1.428571429;
  border-radius: 3px;
  -webkit-transition: all .15s ease-in-out;
  transition: all .15s ease-in-out;
}

.c6:focus {
  outline: 0;
  box-shadow: 0 0 0 3px rgba(34,153,85,0.5);
}

.c6:hover,
.c6:active {
  background-color: #1b7842;
  border-color: #1b7842;
}

.c6[disabled] {
  opacity: .65;
}

.c2 {
  position: relative;
}

.c4 {
  color: #495057;
  margin-bottom: 5px;
  display: none;
}

.c3 {
  margin-bottom: 10px;
}

.c5 {
  display: block;
  width: 100%;
  box-sizing: border-box;
  font-size: 14px;
  line-height: 1.428571429;
  color: #495057;
  border-radius: 3px;
  border-width: 1px;
  border-style: solid;
  border-color: #ced4da;
  background-color: #fff;
  background-image: none;
  padding: 9px 12px;
  -webkit-transition: border-color ease-in-out .15s, box-shadow ease-in-out .15s;
  transition: border-color ease-in-out .15s, box-shadow ease-in-out .15s;
  box-shadow: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c5:focus {
  border-color: #5fdb94;
  box-shadow: 0 0 0 3px rgba(95,219,148,0.5);
  outline: 0;
}

.c5:disabled,
.c5[readonly] {
  background-color: #e9ecef;
}

.c5::-webkit-input-placeholder {
  color: #868e96;
}

.c5::-moz-placeholder {
  color: #868e96;
}

.c5:-ms-input-placeholder {
  color: #868e96;
}

.c5::placeholder {
  color: #868e96;
}

<div>
  <div
    class="c0"
  >
    <div>
      <div
        class="c1"
      >
        passwordless.sms.intro
      </div>
      <form
        class="c2"
        novalidate=""
      >
        <div
          class="c3"
        >
          <label
            class="c4"
            for="phone_number"
          >
            phoneNumber
          </label>
          <input
            autocomplete="tel"
            class="c5"
            data-testid="phone_number"
            id="phone_number"
            name="phone_number"
            placeholder="phoneNumber"
            required=""
            type="tel"
            value=""
          />
        </div>
        <button
          class="c6"
          data-testid="submit"
          type="submit"
        >
          send
        </button>
      </form>
    </div>
  </div>
</div>
`;
